daml-foundations/daml-ghc/ghc-compiler/src/DA/Daml/GHC/Compiler/Convert.hs

```diff
@@ -1022,117 +1022,55 @@ convertCast env expr co = do
-        let lamTo = ETmLam (x, s') (EVar x)
-            lamFrom = ETmLam (x, t') (EVar x)
-        pure (lamTo, lamFrom)
-    | Just (xCo, yCo) <- splitFunCo_maybe co = do
-        let Pair a a' = coercionKind xCo
-        (aTo, aFrom) <- convertCoercion env xCo
-        (bTo, bFrom) <- convertCoercion env yCo
-        sLf <- lift $ convertType env s
-        tLf <- lift $ convertType env t
-        aLf <- lift $ convertType env a
-        a'Lf <- lift $ convertType env a'
-        f <- mkVar <$> mkLamBinder
-        g <- mkVar <$> mkLamBinder
-        x <- mkVar <$> mkLamBinder
-        y <- mkVar <$> mkLamBinder
-        let lamTo =
-                ETmLam
-                    (f, sLf)
-                    (ETmLam
-                         (x, a'Lf)
-                         (bTo `ETmApp` (EVar f `ETmApp` (aFrom `ETmApp` EVar x))))
-        let lamFrom =
-                ETmLam
-                    (g, tLf)
-                    (ETmLam
-                         (y, aLf)
-                         (bFrom `ETmApp` (EVar g `ETmApp` (aTo `ETmApp` EVar y))))
-        pure (lamTo, lamFrom)
-    | Just (aGhc, k_co, co') <- splitForAllCo_maybe co
-    , isReflCo k_co
-     = do
-        let Pair _aK a'K = coercionKind k_co
-        (aTo, aFrom) <- convertCoercion env k_co
-        (bTo, bFrom) <- convertCoercion env co'
-        a'KLf <- lift $ convertKind a'K
-        s' <- lift $ convertType env s
-        t' <- lift $ convertType env t
-        f <- mkVar <$> mkLamBinder
-        g <- mkVar <$> mkLamBinder
-        a' <- mkTypeVar <$> mkLamBinder
-        (a, aKLf) <- lift $ convTypeVar aGhc
-        let lamTo =
-                ETmLam
-                    (f, s')
-                    (ETyLam
-                         (a', a'KLf)
-                         (bTo `ETmApp`
-                          (EVar f `ETmApp` (aFrom `ETyApp` (TVar a')))))
-        let lamFrom =
-                ETmLam
-                    (g, t')
-                    (ETyLam
-                         (a, aKLf)
-                         (bFrom `ETmApp`
-                          (EVar g `ETmApp` (aTo `ETyApp` (TVar a)))))
-        pure (lamTo, lamFrom)
-    -- Case (1) & (2)
-    | Just (tCon, ts, field, flv) <- isSatNewTyCon s t = do
-        newtypeCoercion tCon ts field flv s t
-    | Just (tCon, ts, field, flv) <- isSatNewTyCon t s = do
-        (to, from) <- swap <$> newtypeCoercion tCon ts field flv t s
+convertCoercion :: Env -> Coercion -> StateT Int ConvertM (LF.Expr -> LF.Expr, LF.Expr -> LF.Expr)
+convertCoercion env co@(coercionKind -> Pair s t)
+    | isReflCo co = pure (id, id)
+    | Just (aCo, bCo) <- splitFunCo_maybe co = do
+        let Pair a a' = coercionKind aCo
+        (aTo, aFrom) <- convertCoercion env aCo
+        (bTo, bFrom) <- convertCoercion env bCo
+        a <- lift $ convertType env a
+        a' <- lift $ convertType env a'
+        x <- mkLamBinder
+        let to expr = ETmLam (x, a') $ bTo $ ETmApp expr $ aFrom $ EVar x
```

Comment:
Do you really need fresh variables here? I would have thought capture rules meant that even if they were named identically they wouldn't conflict? That means we can get rid of the StateT, which would simplify things.

Reply:
You're right, we don't necessarily need fresh names here. I'll remove them for now and consider reintroducing them when we have fresh name generation in `ConvertM`.